make torch.load() aware of import path changes

bdhirsh opened the pull request:
pickle.Unpickle provides hooks to let you re-map import statements that are used to unpickle an object. Taking advantage of this to try to preserve BC before landing https://github.com/pytorch/pytorch/pull/51633.

This PR itself doesn't add any new import mappings. But I'm planning on adding one for `torch.tensor` -> `torch._tensor` as part of landing https://github.com/pytorch/pytorch/pull/51633

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53139 make torch.load() aware of import path changes**

Differential Revision: [D26760600](https://our.internmc.facebook.com/intern/diff/D26760600/)

bdhirsh committed:
make torch.load() aware of import path changes: torch.tensor -> torch._tensor



Differential Revision: [D26760600](https://our.internmc.facebook.com/intern/diff/D26760600/)

[ghstack-poisoned]

bdhirsh commented on torch/serialization.py:
```diff
@@ -846,7 +862,11 @@ def persistent_load(saved_id):
     data_file = io.BytesIO(zip_file.get_record(pickle_file))
-    unpickler = pickle_module.Unpickler(data_file, **pickle_load_args)
+
+    if pickle_module.__name__ == 'pickle':
```
This is especially fragile because this fix only works for the `pickle` module. For example, `cpickle` has a slightly different way of doing it.

I opted not to bother with the `cpickle` version because I don't see it currently used anywhere, and didn't want to add it an explicit dependency in this file. But technically, it looks like `torch.load()` lets you pass in an arbitrary module to do the unpickling, as long as it aligns with the `pickle` interface. So https://github.com/pytorch/pytorch/pull/51633 can in theory still break BC if you:
- try to `torch.load()` a tensor that was saved with a version of code that used the `torch.tensor` module
- use `torch.load()` with a custom pickling module

The combination of those two things seems unlikely (hopefully)

ezyang replied:
Can't you do this in a "has-a" way? Your wrapper doesn't inherit from pickle.Unpickler; instead, it forwards all calls on it to the underlying pickler, except find_class which it intercepts to do our custom logic (and if it's a miss, forwards on to the underlying pickler)

bdhirsh replied:
huh I might be misunderstanding- how would you do that in a "has-a" way? Wouldn't you need to explicitly write each function in the Unpickler wrapper that Unpickler.load() normally calls, and write it to forward to the underlying unpickler? It seems like we basically get all of that for free with inheritance (all calls are automatically forwarded to the base Unpickler class, except for `find_class` which we override).

bdhirsh replied:
(from the [docs](https://python.readthedocs.io/en/latest/library/pickle.html#pickle.Unpickler.find_class), it sounds like overriding the class is the standard way to do it, at least for `pickle`. `cpickle` is different though)

ezyang replied:
Oops, I said something that doesn't actually work.

Here is what actually will work though. Look at the existing unpickle implementation, you'll see that we are monkey patching the unpickler object:

```
    unpickler = pickle_module.Unpickler(f, **pickle_load_args)
    unpickler.persistent_load = persistent_load
```

You should just monkey patch `find_class` as well (and have it defer to original, pre-monkey patch `find_class` implementation if there is nothing). This will work universally for both pickle, cpickle and dill.

bdhirsh replied:
Updated the PR and comment got outdated :/

> You should just monkey patch find_class as well
Ah yeah, I actually tried that. It doesn't look like `pickle` lets you do that- I get the error:
```
sts#binary,link-tree/torch/serialization.py", line 875, in _load
    unpickler.find_class = find_class
AttributeError: '_pickle.Unpickler' object attribute 'find_class' is read-only
```

So, a few things I found from looking around more:
- monkey-patching the `find_class` method works seamlessly when I use `dill`, but not for `pickle`. Subclassing appears to work for both though.
- The [stackoverflow](https://stackoverflow.com/questions/13398462/unpickling-python-objects-with-a-changed-module-path/13405732) that I based the changes on makes a distinction between `pickle` and `cpickle`, but it turns out that python3 decides which one to use for you transparently when you import `pickle` (from a stackoverflow [link](https://stackoverflow.com/questions/55406089/how-can-i-know-if-python-3-is-using-cpickle-or-pickle), which points to the implementation [here](https://github.com/python/cpython/blob/d93fbbf88e4abdd24a0a55e3ddf85b8420c62052/Lib/pickle.py#L1598-L1613). I tried to emulate using the non-c implementation by using `pickle._Unpickler` instead of `pickle.Unpickler` (from the [source](https://github.com/python/cpython/blob/d93fbbf88e4abdd24a0a55e3ddf85b8420c62052/Lib/pickle.py#L1598-L1613)), and it looks like subclassing works for both versions.
- In python2 it sounds like `pickle` and `cpickle` had a different API for mapping imports, but it looks to me like they use the same API as part of that change above. From the 

So, it looks like subclassing works in the 3 cases (dill, and both the C and python implementation of pickle). From what I can tell, `dill` and `pickle` are the only modules that we really use for serializing, so I updated the PR to always use the subclass, instead of only doing it for `pickle`. The assumption I'm making is that subclassing on `find_class` is part of the interface of any pickle serialization module- do you think that's reasonable?

I also updated the PR to not actually make the `torch.tensor` -> `torch._tensor` mapping- that needs to land along with the PR that's actually changing the module.

bdhirsh committed:
Update on "make torch.load() aware of import path changes: torch.tensor -> torch._tensor"

pickle.Unpickle provides hooks to let you re-map import statements that are used to unpickle an object. Taking advantage of this to try to preserve BC before landing https://github.com/pytorch/pytorch/pull/51633.

One problem: `torch.load()` accepts an arbitrary module to do the unpickling, but the logic to provide the hooks is module-dependent. I only implemented the logic for the default module, `pickle`.


Differential Revision: [D26760600](https://our.internmc.facebook.com/intern/diff/D26760600/)

[ghstack-poisoned]

bdhirsh committed:
Update on "make torch.load() aware of import path changes: torch.tensor -> torch._tensor"

pickle.Unpickle provides hooks to let you re-map import statements that are used to unpickle an object. Taking advantage of this to try to preserve BC before landing https://github.com/pytorch/pytorch/pull/51633.

One problem: `torch.load()` accepts an arbitrary module to do the unpickling, but the logic to provide the hooks is module-dependent. I only implemented the logic for the default module, `pickle`.


Differential Revision: [D26760600](https://our.internmc.facebook.com/intern/diff/D26760600/)

[ghstack-poisoned]

ezyang commented on torch/serialization.py:
```diff
@@ -860,13 +843,28 @@ def persistent_load(saved_id):
+    # Lets us override the imports that pickle uses when unpickling an object.
+    # This is useful for maintaining BC if we change a module path that tensor instantiation relies on.
+    # NOTE: this only works for the pickle module.
+    # If we switch to a different module (e.g. cpickle), we'll need something else.
```
This comment now out of date?

ezyang commented on torch/serialization.py:
```diff
@@ -844,9 +843,28 @@ def persistent_load(saved_id):
+
+    # Need to subclass Unpickler instead of directly monkey-patching the find_class method
+    # because it's marked readonly in pickle.
+    class UnpicklerWrapper(pickle_module.Unpickler):
```
I'm cool with this

ezyang reviewed (approved):
Don't forget to test, if possible (if only a manual test is convenient, make sure to record it in the Test Plan)

bdhirsh commented:
> Don't forget to test, if possible (if only a manual test is convenient, make sure to record it in the Test Plan)

Yep, I tested manually using a patch of the `tensor`->`_tensor` PR, against pickle and dill. I'll update the test plan (also posted some findings [above](https://github.com/pytorch/pytorch/pull/53139#discussion_r586646062))

It looks like something in the distributed tests doesn't like my change though, looking into that.

bdhirsh committed:
Update on "make torch.load() aware of import path changes"

pickle.Unpickle provides hooks to let you re-map import statements that are used to unpickle an object. Taking advantage of this to try to preserve BC before landing https://github.com/pytorch/pytorch/pull/51633.

This PR itself doesn't add any new import mappings. But I'm planning on adding one for `torch.tensor` -> `torch._tensor` as part of landing https://github.com/pytorch/pytorch/pull/51633


Differential Revision: [D26760600](https://our.internmc.facebook.com/intern/diff/D26760600/)

[ghstack-poisoned]

bdhirsh commented:
@ezyang so, mypy tests are unhappy because I'm dynamically generating new classes and it can't statically determine the type of my `UnpicklerWrapper` class :(

I'm tempted to just add a `# type: ignore` to that line so mypy leaves it alone. Another option would be to write separate versions of that class for both the `pickle` and `dill` modules, since afaik those are the only serialization modules that we care about. Then we can dynamically choose which one to use based on the name of the module that's passed to `torch.load()`. I dumped an example patch at the bottom of this comment. The main downsides:
- This will solidly break if someone ever introduces a new module to do the serialization. The other version will dynamically handle it.
- It adds an explicit `dill` dependency to this file

Then again, a downside of my current PR is that it will generate a new class on the fly every time that `torch.load()` is called. Right now I'm assuming that `torch.load()` never gets called in giant loop or anything. If perf is a concern it might make sense to memoize the classes as they're created.


patch:
```
(/scratch/hirsheybar/pytorch-redispatch_inline-env) hirsheybar@devfair017:/scratch/hirsheybar/pytorch-redispatch_inline$ git diff
diff --git a/torch/serialization.py b/torch/serialization.py
index 214131ab27..42ef725397 100644
--- a/torch/serialization.py
+++ b/torch/serialization.py
@@ -16,6 +16,7 @@ from torch.types import Storage
 from typing import Any, BinaryIO, cast, Dict, Optional, Type, Tuple, Union, IO
 import copyreg
 import pickle
+import dill
 import pathlib

 DEFAULT_PROTOCOL = 2
@@ -818,6 +819,23 @@ def _get_restore_location(map_location):
             return result
     return restore_location

+load_module_mapping: Dict[str, str] = {}
+
+class PickleUnpicklerWrapper(pickle.Unpickler):
+    def find_class(self, mod_name, name):
+        mod_name = load_module_mapping.get(mod_name, mod_name)
+        return super().find_class(mod_name, name)
+
+class DillUnpicklerWrapper(dill.Unpickler):
+    def find_class(self, mod_name, name):
+        mod_name = load_module_mapping.get(mod_name, mod_name)
+        return super().find_class(mod_name, name)
+
+unpickler_wrapper_dict = {
+    'pickle': PickleUnpicklerWrapper,
+    'dill': DillUnpicklerWrapper,
+}
+
 def _load(zip_file, map_location, pickle_module, pickle_file='data.pkl', **pickle_load_args):
     restore_location = _get_restore_location(map_location)

@@ -843,23 +861,11 @@ def _load(zip_file, map_location, pickle_module, pickle_file='data.pkl', **pickl
         storage = loaded_storages[key]
         return storage

-    load_module_mapping = {}
-
-    # Need to subclass Unpickler instead of directly monkey-patching the find_class method
-    # because it's marked readonly in pickle.
-    class UnpicklerWrapper(pickle_module.Unpickler):
-        # from https://stackoverflow.com/questions/13398462/unpickling-python-objects-with-a-changed-module-path/13405732
-        # Lets us override the imports that pickle uses when unpickling an object.
-        # This is useful for maintaining BC if we change a module path that tensor instantiation relies on.
-        def find_class(self, mod_name, name):
-            mod_name = load_module_mapping.get(mod_name, mod_name)
-            return super().find_class(mod_name, name)
-
     # Load the data (which may in turn use `persistent_load` to load tensors)
     data_file = io.BytesIO(zip_file.get_record(pickle_file))

     # unpickler = pickle_module.Unpickler(data_file, **pickle_load_args)
-    unpickler = UnpicklerWrapper(data_file, **pickle_load_args)
+    unpickler = unpickler_wrapper_dict[pickle_module.__name__](data_file, **pickle_load_args)
     unpickler.persistent_load = persistent_load
     result = unpickler.load()

```

ezyang commented:
> I'm tempted to just add a # type: ignore to that line so mypy leaves it alone. 

Just do this.

bdhirsh committed:
Update on "make torch.load() aware of import path changes"

pickle.Unpickle provides hooks to let you re-map import statements that are used to unpickle an object. Taking advantage of this to try to preserve BC before landing https://github.com/pytorch/pytorch/pull/51633.

This PR itself doesn't add any new import mappings. But I'm planning on adding one for `torch.tensor` -> `torch._tensor` as part of landing https://github.com/pytorch/pytorch/pull/51633


Differential Revision: [D26760600](https://our.internmc.facebook.com/intern/diff/D26760600/)

[ghstack-poisoned]

bdhirsh committed:
Update on "make torch.load() aware of import path changes"

pickle.Unpickle provides hooks to let you re-map import statements that are used to unpickle an object. Taking advantage of this to try to preserve BC before landing https://github.com/pytorch/pytorch/pull/51633.

This PR itself doesn't add any new import mappings. But I'm planning on adding one for `torch.tensor` -> `torch._tensor` as part of landing https://github.com/pytorch/pytorch/pull/51633


Differential Revision: [D26760600](https://our.internmc.facebook.com/intern/diff/D26760600/)

[ghstack-poisoned]

bdhirsh committed:
Update on "make torch.load() aware of import path changes"

pickle.Unpickle provides hooks to let you re-map import statements that are used to unpickle an object. Taking advantage of this to try to preserve BC before landing https://github.com/pytorch/pytorch/pull/51633.

This PR itself doesn't add any new import mappings. But I'm planning on adding one for `torch.tensor` -> `torch._tensor` as part of landing https://github.com/pytorch/pytorch/pull/51633


Differential Revision: [D26760600](https://our.internmc.facebook.com/intern/diff/D26760600/)

[ghstack-poisoned]

bdhirsh committed:
Update on "make torch.load() aware of import path changes"

pickle.Unpickle provides hooks to let you re-map import statements that are used to unpickle an object. Taking advantage of this to try to preserve BC before landing https://github.com/pytorch/pytorch/pull/51633.

This PR itself doesn't add any new import mappings. But I'm planning on adding one for `torch.tensor` -> `torch._tensor` as part of landing https://github.com/pytorch/pytorch/pull/51633


Differential Revision: [D26760600](https://our.internmc.facebook.com/intern/diff/D26760600/)

[ghstack-poisoned]

bdhirsh committed:
Update on "make torch.load() aware of import path changes"

pickle.Unpickle provides hooks to let you re-map import statements that are used to unpickle an object. Taking advantage of this to try to preserve BC before landing https://github.com/pytorch/pytorch/pull/51633.

This PR itself doesn't add any new import mappings. But I'm planning on adding one for `torch.tensor` -> `torch._tensor` as part of landing https://github.com/pytorch/pytorch/pull/51633


Differential Revision: [D26760600](https://our.internmc.facebook.com/intern/diff/D26760600/)

[ghstack-poisoned]
